bdk_core/src/sparse_chain.rs

```diff
@@ -350,27 +352,27 @@ impl SparseChain {
 
     pub fn iter_txids(
         &self,
-    ) -> impl DoubleEndedIterator<Item = (TxHeight, Txid)> + ExactSizeIterator + '_ {
-        self.txid_by_height.iter().map(|(k, v)| (*k, *v))
+    ) -> impl DoubleEndedIterator<Item = (TxData<D>, Txid)> + ExactSizeIterator + '_ {
+        self.txid_by_height.iter().map(|(k, v)| (k.clone(), *v))
     }
 
     pub fn iter_txids_in_height_range<R: RangeBounds<TxHeight>>(
         &self,
         range: R,
     ) -> impl DoubleEndedIterator + '_ {
-        fn map_bound(bound: Bound<&TxHeight>) -> Bound<(TxHeight, Txid)> {
+        fn map_bound<D: Clone + Default>(bound: Bound<&TxHeight>) -> Bound<(TxData<D>, Txid)> {
             match bound {
                 Bound::Unbounded => Bound::Unbounded,
-                Bound::Included(x) => Bound::Included((x.clone(), Txid::all_zeros())),
-                Bound::Excluded(x) => Bound::Excluded((x.clone(), Txid::all_zeros())),
+                Bound::Included(x) => Bound::Included((x.clone().into(), Txid::all_zeros())),
+                Bound::Excluded(x) => Bound::Excluded((x.clone().into(), Txid::all_zeros())),
```

Comment:
Good point, how about:

```rust
pub struct TxData<D> {
    pub height: TxHeight,
    pub data: Option<D>,
}
```

This way when `data` is none, it will always be smaller than when data is `Some`.

Reply:
I'm not sure that solves the problem. You still need to find the `max` of `Option<D>` to put it in the bounds. I think you want to people to be able to query on the `D`. I mean that's why you've put it in this index right?

We have to decide whether the data is meant to be changing the order of stuff (in which case it should be included in the range) or whether it's just additional data in which case we don't put it in the height index.

Reply:
I feel like querying by `TxHeight` should be enough for now? As long as we have `min` for `Option<D>` we can query by `TxHeight` only.

Reply:
Fixed here:

https://github.com/LLFourn/bdk_core_staging/blob/430c2cf7478663cabf5c2d984c5c16b5fdda51a7/bdk_core/src/sparse_chain.rs#L392-L424